src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorFileChangeDetector.cs

```diff
@@ -130,11 +145,69 @@ protected virtual IReadOnlyList<string> GetExistingRazorFiles(string workspaceDi
             return existingRazorFiles.ToArray();
         }
 
-        private void FileSystemWatcher_RazorFileEvent_Background(string physicalFilePath, RazorFileChangeKind kind)
+        // Internal for testing
+        internal void FileSystemWatcher_RazorFileEvent_Background(string physicalFilePath, RazorFileChangeKind kind)
         {
-            Task.Factory.StartNew(
-                () => FileSystemWatcher_RazorFileEvent(physicalFilePath, kind),
-                CancellationToken.None, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            lock (_pendingNotificationsLock)
+            {
+                if (!_pendingNotifications.TryGetValue(physicalFilePath, out var currentNotification))
+                {
+                    currentNotification = new DelayedFileChangeNotification();
+                    _pendingNotifications[physicalFilePath] = currentNotification;
+                }
+
+                if (currentNotification.ChangeKind != null)
+                {
+                    Debug.Assert(currentNotification.ChangeKind == RazorFileChangeKind.Added || currentNotification.ChangeKind == RazorFileChangeKind.Removed);
```

Comment:
Why can't this be `Changed` kind?

Reply:
Not sure I understand the question. Are you asking why is it `currentNotification.ChangeKind`?

Reply:
Why can't the ChangeKind be `RazorFileChangeKind.Changed`?

Reply:
oo because we don't hook into that event in the filewatchers above.